106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,583 @@
+  * **Idle**: if a scale up operation was requested, it was executed and completed successfully/failed or a full rebalance was asked due to an anomaly but since the scale-down rebalance is done, we can ignore the anomalies assuming they are fixed by the rebalance. In case, they are not fixed, Cruise Control will detect them again and a new rebalance would be requested.
+
+```mermaid
+sequenceDiagram
+  autonumber
+  RebalanceOnImbalance->>RebalanceOnScaleDown: requested scale down
+  RebalanceOnImbalance->>RebalanceOnScaleUp: requested scale up
+  RebalanceOnImbalance->>RebalanceOnImbalance: new full rebalance requested
+  critical queued scale up and skew rebalance
+    RebalanceOnImbalance->>RebalanceOnScaleUp: execute scale up
+    RebalanceOnScaleUp-->>RebalanceOnImbalance: then full rebalance
+  end
+  critical queued scale down and skew rebalance
+    RebalanceOnImbalance->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnImbalance: then full rebalance
+  end
+  critical queued scale down and scale up and skew rebalance
+    RebalanceOnImbalance->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnImbalance: then full rebalance
+  end
+  RebalanceOnImbalance->>Idle: scale up complete with no queued items
+```
+* from **RebalanceOnImbalance**:
+  * **RebalanceOnImbalance**: if another goal violation was detected while the first one ended. If a scale down and scale up is also queued up then they will execute first.
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is queued and will run if there is no other rebalancing scale down in queue. If a rebalancing scale down is in queue then it will be executed first.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, it will run once the full rebalance is completed
+  * **Idle**: if full rebalance was requested, it was executed and completed successfully or failed.
+
+On each reconciliation, the following process will be used:
+
+```mermaid
+flowchart TB
+  A[KafkaClusterCreator] --creates--> B[KafkaCluster]
+  B -- calls --> D{Reconcilation in Progress}
+  D -- no --> E{if config map present?}
+  D -- yes --> F[fix the anomalies]
+  E -- yes --> G[Trigger auto-rebalance]
+  E -- no --> H[No operation]
+```
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance.
+2. The `KafkaAutoRebalancingReconciler.reconcile()` will then check if there was any ConfigMap created with name  `goal-violation-map` and whether the `anomalyList` list is empty or not, then the `full` rebalance(imbalance mode) would be performed if the list is not empty.
+3. If a rebalance is already ongoing and more anomalies are detected, then the operator will just ignore the new anomalies based on the timestamps as discussed earlier in the proposal and delete all the anomalies from the `anomalyList` in the ConfigMap.
```

Comment:
I don't think these description matches the diagram very well. I understand the diagram is high level and then you have low level descriptions here, maybe that makes it confusing.